forge/src/runner.rs

```diff
@@ -210,251 +186,234 @@ impl<'a, DB: DatabaseRef + Send + Sync> ContractRunner<'a, DB> {
+                setup.logs.extend(execution_logs);
+                (status, None, gas_used, execution_trace, state_changeset)
+            }
+            Err(EvmError::Execution {
+                status,
+                reason,
+                gas_used,
+                logs: execution_logs,
+                traces: execution_trace,
+                labels: new_labels,
+                state_changeset,
+                ..
+            }) => {
+                setup.labeled_addresses.extend(new_labels);
+                setup.logs.extend(execution_logs);
+                (status, Some(reason), gas_used, execution_trace, state_changeset)
+            }
+            Err(err) => {
+                tracing::error!(?err);
+                return Err(err.into())
             }
-            Err(err) => match err {
-                EvmError::Execution {
-                    status,
-                    reason,
-                    gas_used,
-                    logs: execution_logs,
-                    state_changeset,
-                } => {
-                    logs.extend(execution_logs);
-                    (status, Some(reason), gas_used, logs, state_changeset)
-                }
-                err => {
-                    tracing::error!(?err);
-                    return Err(err.into())
-                }
-            },
         };
+        setup
+            .traces
+            .extend(execution_traces.map(|traces| (TraceKind::Execution, traces)).into_iter());
 
-        // DSTest will not revert inside its `assertEq`-like functions
-        // which allows to test multiple assertions in 1 test function while also
-        // preserving logs - instead it sets `failed` to `true` which we must check.
         let success = self.executor.is_success(
-            address,
+            setup.address,
             status,
             state_changeset.expect("we should have a state changeset"),
             should_fail,
         );
-        let duration = Instant::now().duration_since(start);
-        tracing::debug!(?duration, %success, %gas_used);
+
+        // Record test execution time
+        tracing::debug!(
+            duration = ?Instant::now().duration_since(start),
+            %success,
+            %gas_used
+        );
 
         Ok(TestResult {
             success,
             reason,
-            gas_used,
             counterexample: None,
-            logs,
+            logs: setup.logs,
             kind: TestKind::Standard(gas_used),
-            traces: None,
-            identified_contracts,
+            traces: setup.traces,
             debug_calls: None,
-            labeled_addresses: Default::default(),
+            labeled_addresses: setup.labeled_addresses,
         })
     }
 
-    #[tracing::instrument(name = "fuzz-test", skip_all, fields(name = %func.signature()))]
+    #[tracing::instrument(name = "fuzz-test", skip_all, fields(name = %func.signature(), %should_fail))]
     pub fn run_fuzz_test(
```

Comment:
This function is a lot simpler since the final "rerun failed fuzz case" step is in `FuzzedExecutor` now. The return type if `FuzzedExecutor::fuzz` also does not leak the proptest error, so we don't need to do complicated matching to figure out if we failed and why (this is also handled in `FuzzedExecutor`)